msprime/species_tree.py

```diff
@@ -0,0 +1,834 @@
+                            root_edge_found = True
+                            new_length = edge.get_length()
+                            new_length += max_total_edge_length
+                            new_length -= total_edge_length
+                            edge.set_length(round(new_length, 8))
+                        else:
+                            this_edge = parent_edge
+        # First specify the edges for which the parents still need to
+        # be identified.
+        for edge in self.get_edges():
+            if edge.get_is_root_edge():
+                edge.set_parent_needs_times(False)
+            else:
+                edge.set_parent_needs_times(True)
+        # Set the times of all edges.
+        for edge in self.get_edges():
+            if edge.get_is_extant() is True:
+                edge.set_termination(0.0)
+                edge.set_origin(edge.get_length())
+                this_edge = edge
+                while this_edge.get_parent_needs_times():
+                    for other_edge in self.get_edges():
+                        if other_edge.get_node_ids()[1] == this_edge.get_node_ids()[0]:
+                            parent_edge = other_edge
+                            break
+                    err = 'The parent edge for edge '
+                    err += '{} could not be found'.format(this_edge.get_id())
+                    assert parent_edge, err
+                    parent_edge.set_termination(this_edge.get_origin())
+                    parent_edge.set_origin(this_edge.get_origin() +
+                                           parent_edge.get_length())
+                    this_edge.set_parent_needs_times = False
+                    this_edge = parent_edge
+
+    def get_origin(self):
+        for edge in self.get_edges():
+            if edge.get_is_root_edge():
+                return edge.get_origin()
+
+    def info(self):
+        info_string = ''
+        info_string += 'Tree'.ljust(20)
+        info_string += '\n'
+        info_string += 'Number of edges:'.ljust(20)
+        info_string += str(self.get_number_of_edges())
+        info_string += '\n'
+        return info_string
+
+
+class Edge(object):
+    """
+    The edge object, with two IDs corresponding to the nodes at both
+    ends of the edge, and ages of origin and termination.
+    """
+    def __init__(self, id):
+        self.id = id
+        self.node_ids = []
+        self.extant_progeny_ids = []
+        self.origin = None
+        self.termination = None
+        self.dmv = None
+        self.parent_needs_times = True
+
+    def get_id(self):
+        return self.id
+
+    def set_node_ids(self, node_ids):
+        self.node_ids = node_ids
+
+    def get_node_ids(self):
+        return self.node_ids
+
+    def get_is_extant(self):
+        if self.node_ids[1][0:12] == 'internalNode':
+            return False
+        else:
+            return True
+
+    def set_length(self, length):
```

Comment:
I don't think we need all the Java-style getters and setters here, we can just access and set the ``length`` attribute directly (particularly since this is an internal API, and won't be exposed to the user)

Reply:
Ok, thanks. All these are no longer included.